File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/kakurasu/domain/KakurasuGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.kakurasu.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.Arrays;

import fj.F;

/**
 * Created by zwvista on 2016/09/29.
 */

public class KakurasuGameState extends CellsGameState<KakurasuGame, KakurasuGameMove, KakurasuGameState> {
    public KakurasuObject[] objArray;
    public HintState[] row2state;
    public HintState[] col2state;

    public KakurasuGameState(KakurasuGame game) {
        super(game);
        objArray = new KakurasuObject[rows() * cols()];
        Arrays.fill(objArray, KakurasuObject.Empty);
        row2state = new HintState[rows() * 2];
        col2state = new HintState[cols() * 2];
        updateIsSolved();
    }

    public KakurasuObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public KakurasuObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, KakurasuObject obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, KakurasuObject obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(KakurasuGameMove move) {
        Position p = move.p;
        if (!isValid(p) || get(p) == move.obj) return false;
        set(p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(KakurasuGameMove move) {
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<KakurasuObject, KakurasuObject> f = obj -> {
            switch (obj) {
            case Empty:
                return markerOption == MarkerOptions.MarkerFirst ?
                        KakurasuObject.Marker : KakurasuObject.Cloud;
            case Cloud:
                return markerOption == MarkerOptions.MarkerLast ?
                        KakurasuObject.Marker : KakurasuObject.Empty;
            case Marker:
                return markerOption == MarkerOptions.MarkerFirst ?
                        KakurasuObject.Cloud : KakurasuObject.Empty;
            }
            return obj;
        };
        Position p = move.p;
        if (!isValid(p)) return false;
        move.obj = f.f(get(p));
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 8/Kakurasu

        Summary
        Cloud Kakuro on a Skyscraper

        Description
        1. On the bottom and right border, you see the value of (respectively)
           the columns and rows.
        2. On the other borders, on the top and the left, you see the hints about
           which tile have to be filled on the board. These numbers represent the
           sum of the values mentioned above.
    */
    private void updateIsSolved() {
        boolean allowedObjectsOnly = game.gdi.isAllowedObjectsOnly();
        isSolved = true;
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++)
                if (get(r, c) == KakurasuObject.Forbidden)
                    set(r, c, KakurasuObject.Empty);
        for (int r = 1; r < rows() - 1; r++) {
            // 1. On the bottom and right border, you see the value of (respectively)
            // the rows.
            int n1 = 0, n2 = game.row2hint[r * 2];
            for (int c = 1; c < cols() - 1; c++)
                if (get(r, c) == KakurasuObject.Cloud)
                    // 2. On the other borders, on the top and the left, you see the hints about
                    // which tile have to be filled on the board.
                    n1 += game.col2hint[c * 2 + 1];
            // 2. These numbers represent the sum of the values mentioned above.
            HintState s = n1 == 0 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            row2state[r * 2] = s;
            if (s != HintState.Complete) isSolved = false;
            if (n1 >= n2 && allowedObjectsOnly)
                for (int c = 1; c < cols() - 1; c++)
                    switch (get(r, c)) {
                    case Empty: case Marker: set(r, c, KakurasuObject.Forbidden); break;
                    }
        }
        for (int c = 1; c < cols() - 1; c++) {
            // 1. On the bottom and right border, you see the value of (respectively)
            // the columns.
            int n1 = 0, n2 = game.col2hint[c * 2];
            for (int r = 1; r < rows() - 1; r++)
                if (get(r, c) == KakurasuObject.Cloud)
                    // 2. On the other borders, on the top and the left, you see the hints about
                    // which tile have to be filled on the board.
                    n1 += game.row2hint[r * 2 + 1];
            // 2. These numbers represent the sum of the values mentioned above.
            HintState s = n1 == 0 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            col2state[c * 2] = s;
            if (s != HintState.Complete) isSolved = false;
            if (n1 >= n2 && allowedObjectsOnly)
                for (int r = 1; r < rows() - 1; r++)
                    switch (get(r, c)) {
                    case Empty: case Marker: set(r, c, KakurasuObject.Forbidden); break;
                    }
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/kakurasu/domain/KakurasuGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.kakurasu.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState
import fj.F
import java.util.*

class KakurasuGameState(game: KakurasuGame?) : CellsGameState<KakurasuGame?, KakurasuGameMove?, KakurasuGameState?>(game) {
    var objArray: Array<KakurasuObject?>
    var row2state: Array<HintState?>
    var col2state: Array<HintState?>
    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]

    operator fun get(p: Position) = get(p!!.row, p.col)

    operator fun set(row: Int, col: Int, obj: KakurasuObject?) {
        objArray[row * cols() + col] = obj
    }

    operator fun set(p: Position, obj: KakurasuObject?) {
        set(p!!.row, p.col, obj)
    }

    fun setObject(move: KakurasuGameMove): Boolean {
        val p = move!!.p
        if (!isValid(p) || get(p) == move.obj) return false
        set(p, move.obj)
        updateIsSolved()
        return true
    }

    fun switchObject(move: KakurasuGameMove): Boolean {
        val markerOption = MarkerOptions.values()[game!!.gdi.markerOption]
        val f = label@ F { obj: KakurasuObject? ->
            when (obj) {
                KakurasuObject.Empty -> return@label if (markerOption == MarkerOptions.MarkerFirst) KakurasuObject.Marker else KakurasuObject.Cloud
                KakurasuObject.Cloud -> return@label if (markerOption == MarkerOptions.MarkerLast) KakurasuObject.Marker else KakurasuObject.Empty
                KakurasuObject.Marker -> return@label if (markerOption == MarkerOptions.MarkerFirst) KakurasuObject.Cloud else KakurasuObject.Empty
            }
            obj
        }
        val p = move!!.p
        if (!isValid(p)) return false
        move.obj = f.f(get(p))
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 8/Kakurasu

        Summary
        Cloud Kakuro on a Skyscraper

        Description
        1. On the bottom and right border, you see the value of (respectively)
           the columns and rows.
        2. On the other borders, on the top and the left, you see the hints about
           which tile have to be filled on the board. These numbers represent the
           sum of the values mentioned above.
    */
    private fun updateIsSolved() {
        val allowedObjectsOnly = game!!.gdi.isAllowedObjectsOnly
        isSolved = true
        for (r in 0 until rows()) for (c in 0 until cols()) if (get(r, c) == KakurasuObject.Forbidden) set(r, c, KakurasuObject.Empty)
        for (r in 1 until rows() - 1) {
            // 1. On the bottom and right border, you see the value of (respectively)
            // the rows.
            var n1 = 0
            val n2 = game!!.row2hint[r * 2]
            for (c in 1 until cols() - 1) if (get(r, c) == KakurasuObject.Cloud) // 2. On the other borders, on the top and the left, you see the hints about
            // which tile have to be filled on the board.
                n1 += game!!.col2hint[c * 2 + 1]
            // 2. These numbers represent the sum of the values mentioned above.
            val s = if (n1 == 0) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            row2state[r * 2] = s
            if (s != HintState.Complete) isSolved = false
            if (n1 >= n2 && allowedObjectsOnly) for (c in 1 until cols() - 1) when (get(r, c)) {
                KakurasuObject.Empty, KakurasuObject.Marker -> set(r, c, KakurasuObject.Forbidden)
            }
        }
        for (c in 1 until cols() - 1) {
            // 1. On the bottom and right border, you see the value of (respectively)
            // the columns.
            var n1 = 0
            val n2 = game!!.col2hint[c * 2]
            for (r in 1 until rows() - 1) if (get(r, c) == KakurasuObject.Cloud) // 2. On the other borders, on the top and the left, you see the hints about
            // which tile have to be filled on the board.
                n1 += game!!.row2hint[r * 2 + 1]
            // 2. These numbers represent the sum of the values mentioned above.
            val s = if (n1 == 0) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            col2state[c * 2] = s
            if (s != HintState.Complete) isSolved = false
            if (n1 >= n2 && allowedObjectsOnly) for (r in 1 until rows() - 1) when (get(r, c)) {
                KakurasuObject.Empty, KakurasuObject.Marker -> set(r, c, KakurasuObject.Forbidden)
            }
        }
    }

    init {
        objArray = arrayOfNulls(rows() * cols())
        Arrays.fill(objArray, KakurasuObject.Empty)
        row2state = arrayOfNulls(rows() * 2)
        col2state = arrayOfNulls(cols() * 2)
        updateIsSolved()
    }
}